app/src/main/java/com/sdp13epfl2021/projmag/database/impl/firebase/FirebaseCandidatureDatabase.kt

```diff
@@ -81,45 +78,38 @@ class FirebaseCandidatureDatabase(
         userID: String,
         state: Candidature.State
     ) {
-        var waiting = true
-        var profile: ImmutableProfile? = null
-        var cv: CurriculumVitae? = null
-        /*userdataDatabase.getProfile(userID, { profile = it }, { waiting = false})*/ //TODO add when implemented
-        /*userdataDatabase.getCV(userID, { cv = it }, { waiting = false })*/ //TODO add when implemented
-        profile = dummyProfile(userID)
-        cv = dummyCV(userID)
-
-        while (waiting && (profile == null || cv == null)) {
+        val waiting = AtomicBoolean(true)
+        val profileDeferred: CompletableDeferred<ImmutableProfile?> = CompletableDeferred()
+        val cvDeferred: CompletableDeferred<CurriculumVitae?> = CompletableDeferred()
+
+        userdataDatabase.getProfile(userID, { profile ->
+            profileDeferred.complete(profile)
+            // If the profile is null, we stop to wait because no candidature will be added
+            if (profile == null) {
+                waiting.set(false)
+            }
+        }, { waiting.set(false) }) // If a problem occurs, we stop to wait because no candidature will be added
+
+        userdataDatabase.getCv(userID, { cv ->
+            cvDeferred.complete(cv)
+            // If the cv is null, we stop to wait because no candidature will be added
```

Comment:
so do we consider that there needs to already be a profile and cv for a student to apply for a project (i.e. userId alone is not sufficient)?

Reply:
getListOfCandidatures will only return a list of complet candidature.
If a student apply to a project, normally its profile/cv should already exists.